File: coding_py/boj/bj2580.py
a = [list(map(int, input().split())) for _ in range(9)]
b, n = [0]*81, 0
row = [[False]*10 for _ in range(9)]
col = [[False]*10 for _ in range(9)]
squ = [[False]*10 for _ in range(9)]

def s(i, j):
    return i//3*3 + j//3

def solve(idx):
    if idx == n:
        for i in range(9):
            print(' '.join(map(str, a[i])))
        exit(0)
    x, y = b[idx]//9, b[idx]%9
    for i in range(1, 10):
        if not (row[x][i] or col[y][i] or squ[s(x,y)][i]):
            row[x][i] = col[y][i] = squ[s(x,y)][i] = True
            a[x][y] = i
            solve(idx+1)
            a[x][y] = 0
            row[x][i] = col[y][i] = squ[s(x,y)][i] = False

for i in range(9):
    for j in range(9):
        k = a[i][j]
        if k:
            row[i][k] = True
            col[j][k] = True
            squ[s(i,j)][k] = True
        else:
            b[n] = i*9 + j
            n += 1
solve(0)

File: algo_cpp/bj2580.cpp
#include <vector>
#include <iostream>
using namespace std;

vector<pair<int, int>> f;
int cnt;
bool row[9][9];
bool col[9][9];
bool sqr[9][9];
int sudoku[9][9];

void dfs(int idx) {
	if (idx == cnt) {
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				cout << sudoku[i][j]<<' ';
			}
			cout << '\n';
		}
		exit(0);
	}
	int x, y;
	x = f[idx].first;
	y = f[idx].second;
	for (int i = 0; i < 9; i++) {
		if (!row[x][i] && !col[y][i] && !sqr[x / 3 * 3 + y / 3][i]) {
			row[x][i] = col[y][i] = sqr[x / 3 * 3 + y / 3][i] = 1;
			sudoku[x][y] = i + 1;
			dfs(idx + 1);
			sudoku[x][y] = row[x][i] = col[y][i] = sqr[x / 3 * 3 + y / 3][i] = 0;
		}
	}
}

int main(void) {
	for (int i = 0; i < 9; i++) {
		for (int j = 0; j < 9; j++) {
			cin >> sudoku[i][j];
			if (!sudoku[i][j])
				f.push_back({ i,j });
			else
			{
				row[i][sudoku[i][j] - 1] = 1;
				col[j][sudoku[i][j] - 1] = 1;
				sqr[i/3*3+j/3][sudoku[i][j] - 1] = 1;
			}
		}
	}
	cnt = f.size();
	dfs(0);
	return 0;
}